### PR TITLE
Self Removal: Send LeaveRequest Message

### PR DESCRIPTION
**Pull request by mchenani**
### Add group self-removal by sending `xmtp_mls::Group::leave_group` LeaveRequest message for FFI, Node, WASM, content types, storage, and error handling
This pull request introduces a leave request flow across the group stack, adding a `LeaveRequest` content type and a `GroupMembershipState::PendingRemove` state, a `LeaveRequestCodec`, and a `leave_group` operation with validation and message sending. It adds FFI, Node, and WASM bindings for the new state and leave action, database mappings for persistence, and structured errors for disallowed scenarios, along with tests validating behavior. • Implement `xmtp_mls::Group::leave_group` with membership and role validations and sending a `LeaveRequest` via `LeaveRequestCodec` in [mod.rs](https://github.com/xmtp/libxmtp/pull/2606/files#diff-c29f56a38916c7410eff8091df1a2e43487ffe20646d96827e846e475f4608d3) • Expose `FfiConversation::leave_group` to FFI consumers in [mls.rs](https://github.com/xmtp/libxmtp/pull/2606/files#diff-3a24c3e76565487a710ac9863ac05160128f4f90892e07849b555a6de43a6e8f) • Add `LeaveRequest` content type and codec with encode/decode in [leave_request.rs](https://github.com/xmtp/libxmtp/pull/2606/files#diff-67d8cf386920d12a812ecb868102e2c1c873c3fcc8f2e3925831ed8af0e852d8) • Add `PendingRemove` to `GroupMembershipState` in bindings and DB, and map conversions in [group.rs](https://github.com/xmtp/libxmtp/pull/2606/files#diff-e3d4940853a1045db514d5c1a48581d9ef53da4ba45a7cb8387c72c99c475b5a) and [convert.rs](https://github.com/xmtp/libxmtp/pull/2606/files#diff-80e9660ae1cb1f185062680fa2404c845b38e5afb66b3853a190d28b76721378) • Support storing and displaying `LeaveRequest` in group messages in [group_message.rs](https://github.com/xmtp/libxmtp/pull/2606/files#diff-f81cb97873c6ce33f85646f59ec1ea72f8574b06550596ec29a4ac58c2353373) • Add Node/WASM bindings for `ContentType::LeaveRequest` and `GroupMembershipState::PendingRemove` • Add leave-related errors in [error.rs](https://github.com/xmtp/libxmtp/pull/2606/files#diff-a74b7b67dfb7042e191e448698c992366d2c942fcded39848cdb93e50e1439a5) and tests in [mod.rs](https://github.com/xmtp/libxmtp/pull/2606/files#diff-3943e92f9d0be4ce83280dcbb75f2f222fc96342e69cb677bcd2b006e2182c4a)

#### 📍Where to Start
Start with the `xmtp_mls::Group::leave_group` implementation and validations in [mod.rs](https://github.com/xmtp/libxmtp/pull/2606/files#diff-c29f56a38916c7410eff8091df1a2e43487ffe20646d96827e846e475f4608d3), then review the `LeaveRequestCodec` in [leave_request.rs](https://github.com/xmtp/libxmtp/pull/2606/files#diff-67d8cf386920d12a812ecb868102e2c1c873c3fcc8f2e3925831ed8af0e852d8) and the DB content type mappings in [group_message.rs](https://github.com/xmtp/libxmtp/pull/2606/files#diff-f81cb97873c6ce33f85646f59ec1ea72f8574b06550596ec29a4ac58c2353373).

#### Changes since #2606 opened

- Implemented core self-removal message flow with LeaveRequest processing, pending removal tracking, and membership state management [3736c06]
- Created database persistence layer for pending removals and group leave request status [3736c06]
- Implemented background worker and admin methods for processing pending self-removals [3736c06]
- Updated conversion utilities and FFI bindings to expose left_inboxes for self-removal events [3736c06]
- Added comprehensive test coverage for self-removal workflows and edge cases [3736c06]
----

<details>
<summary>📊 <a href="https://app.macroscope.com">Macroscope</a> summarized 3736c06. 21 files reviewed, 3 issues evaluated, 3 issues filtered, 0 comments posted</summary>

### 🗂️ Filtered Issues
<details>
<summary>xmtp_db/src/encrypted_store/group.rs — 0 comments posted, 1 evaluated, 1 filtered</summary>

- [line 1174](https://github.com/xmtp/libxmtp/blob/3736c06ca6546db8c5bd89f876411d7e99937ece/xmtp_db/src/encrypted_store/group.rs#L1174): Cross-layer enum discriminant mismatch for `GroupMembershipState::PendingRemove` likely causes wrong state mapping across FFI/Node/WASM/DB boundaries. In `xmtp_db/src/encrypted_store/group.rs`, `GroupMembershipState::PendingRemove = 5` (id 21), but in Node bindings (`bindings_node/src/conversations.rs`, ids 4–6) and WASM bindings (`bindings_wasm/src/conversations.rs`, ids 7–9), `PendingRemove = 4`. If these enums are serialized/deserialized as integers across boundaries (which is typical for FFI and JS/WASM bindings) or persisted as integers, the mismatch will produce incorrect interpretation of states (e.g., a value 4 coming from Node/WASM would not map to `PendingRemove` in Rust/DB, and a 5 from DB would not map to `PendingRemove` in Node/WASM). This is a runtime contract bug causing visible incorrect behavior rather than a compile error. <b>[ Low confidence ]</b>
</details>

<details>
<summary>xmtp_db/src/encrypted_store/group_message.rs — 0 comments posted, 1 evaluated, 1 filtered</summary>

- [line 146](https://github.com/xmtp/libxmtp/blob/3736c06ca6546db8c5bd89f876411d7e99937ece/xmtp_db/src/encrypted_store/group_message.rs#L146): Potential contract parity gap: `LeaveRequest` content type added but codec definition may be incomplete or absent, leading to runtime handling gaps. Several places assume a `leave_request` content type exists: - `xmtp_db/src/encrypted_store/group_message.rs` adds `ContentType::LeaveRequest` (id 28) with `Display` and string parsing mapping to `leave_request::LeaveRequestCodec::TYPE_ID` (ids 30–31). - `xmtp_mls/src/groups/mls_sync.rs` adds processing for external messages when `message.content_type == ContentType::LeaveRequest` (id 58). - `xmtp_content_types/src/lib.rs` and `bindings_node` add `LeaveRequest` enum variants (ids 19 and 2–3). However, `xmtp_content_types/src/leave_request.rs` appears to be empty or placeholder-only (ids 10–18), and the `LeaveRequestCodec` type and `TYPE_ID` must exist and be wired into the registry for encode/decode. If the codec is not fully implemented and registered, messages of this type may not be properly serialized/deserialized at runtime, causing them to be treated as unknown or leading to no-ops in processing. This is a runtime behavior bug (silently unhandled content) rather than a compile error only if `LeaveRequestCodec` compiles but is not properly registered or returns errors at runtime. <b>[ Low confidence ]</b>
</details>

<details>
<summary>xmtp_mls/src/groups/mls_sync.rs — 0 comments posted, 1 evaluated, 1 filtered</summary>

- [line 701](https://github.com/xmtp/libxmtp/blob/3736c06ca6546db8c5bd89f876411d7e99937ece/xmtp_mls/src/groups/mls_sync.rs#L701): Order-of-operations bug causes loss of "left vs removed" classification in `GroupUpdated` payloads. In `xmtp_mls/src/groups/mls_sync.rs`, `clean_pending_remove_list(storage, &validated_commit.removed_inboxes)` is called before building the `GroupUpdated` payload using `validated_commit.into_with(pending_remove_users)` (ids 61 and 63). The `into_with` implementation (id 89) classifies `removed_inboxes` into `left_inboxes` vs `removed_inboxes` based on the current `pending_remove_users` fetched from storage. Because the cleanup runs before payload construction, any entries corresponding to members who requested self-removal will be removed from the pending list, causing them to be misclassified as `removed_inboxes` instead of `left_inboxes`. This results in incorrect events/logging and FFI exposure (`FfiGroupUpdated.left_inboxes`) always being empty for those cases. <b>[ Low confidence ]</b>
</details>

</details>

### CODE REVIEW

**Comment by mchenani** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/xmtp/libxmtp/2606?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#2611** <a href="https://app.graphite.dev/github/pr/xmtp/libxmtp/2611?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2610** <a href="https://app.graphite.dev/github/pr/xmtp/libxmtp/2610?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2608** <a href="https://app.graphite.dev/github/pr/xmtp/libxmtp/2608?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2606** <a href="https://app.graphite.dev/github/pr/xmtp/libxmtp/2606?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/xmtp/libxmtp/2606?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#2605** <a href="https://app.graphite.dev/github/pr/xmtp/libxmtp/2605?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>mergequeue</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/xmtp?ref=merge-queue-instructions-in-stack-comment&prId=2606).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: xmtp_mls/src/groups/error.rs [hunk 1]
```diff
@@ -208,6 +210,32 @@ impl RetryableError for MetadataPermissionsError {
...
 }
 
+#[derive(Error, Debug)]
+pub enum GroupLeaveValidationError {
+    #[error("cannot leave a DM conversation")]
+    DmLeaveForbidden,
+    #[error("cannot leave a group that has only one member")]
+    SingleMemberLeaveRejected,
+    #[error("cannot leave a group without an admin or super admin")]
+    LeaveWithoutAdminForbidden,
```
Comment (mchenani) [inline xmtp_mls/src/groups/error.rs]: change to LeaveWithoutSuperAdminForbidden

File: xmtp_mls/src/groups/error.rs [hunk 2]
```diff
@@ -208,6 +210,32 @@ impl RetryableError for MetadataPermissionsError {
...
+    #[error("inbox ID already exists in the pending leave list")]
+    InboxAlreadyInPendingList,
+    #[error("inbox ID does not exist in the pending leave list")]
+    InboxNotInPendingList,
+    #[error("inbox ID must be removed from group members before leaving")]
+    MustRemoveFromMembersFirst,
```
Comment (mchenani) [inline xmtp_mls/src/groups/error.rs]: remove this

File: xmtp_mls/src/groups/error.rs [hunk 3]
```diff
@@ -208,6 +210,32 @@ impl RetryableError for MetadataPermissionsError {
...
+    #[error("only the member themselves can retract a pending leave request")]
+    OnlyMemberCanRetractLeave,
```
Comment (mchenani) [inline xmtp_mls/src/groups/error.rs]: remove this

File: xmtp_mls/src/groups/mod.rs [hunk 4]
```diff
@@ -979,6 +982,57 @@ where
...
+            return Err(GroupLeaveValidationError::DmLeaveForbidden.into());
+        }
+
+        let is_admin = self.is_admin(self.context.inbox_id().to_string())?;
+        let is_super_admin = self.is_super_admin(self.context.inbox_id().to_string())?;
+        let admin_size = self.admin_list()?.len();
+        let super_admin_size = self.super_admin_list()?.len();
+
+        // check if the user is the only Admin or SuperAdmin of the group
+        if (is_admin && admin_size == 1) || (is_super_admin && super_admin_size == 1) {
```
Comment (mchenani) [inline xmtp_mls/src/groups/mod.rs]: consider to check if it's admin they have the permission or not

**Review by cameronvoell — APPROVED** [review]
